From: =?utf-8?b?Ik1hcmNvIFRyZXZpc2FuIChUcmV2acOxbyki?= <mail@3v1n0.net>
Date: Wed, 1 May 2024 01:14:43 +0200
Subject: js: Use Shell.util_spawn_async functions to launch external
 processes

As explained in previous commits, it's not safe to use JS code in child
function callbacks, so let's use the safer version of it.

Fixes: https://gitlab.gnome.org/GNOME/gnome-shell/-/issues/6698
Part-of: <https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/3293>
(cherry picked from commit 26e8fb90fb595510240c57b0862013dbf4739770)
(cherry picked from commit e633b1d4f59723d11273eebb8cf2043ba2e44713)
---
 js/misc/ibusManager.js           | 11 ++---------
 js/misc/util.js                  | 11 ++---------
 js/ui/components/networkAgent.js | 12 +++---------
 3 files changed, 7 insertions(+), 27 deletions(-)

diff --git a/js/misc/ibusManager.js b/js/misc/ibusManager.js
index c3022aa..6a51587 100644
--- a/js/misc/ibusManager.js
+++ b/js/misc/ibusManager.js
@@ -115,16 +115,9 @@ class IBusManager extends Signals.EventEmitter {
             const env = launchContext.get_environment();
             // Use DO_NOT_REAP_CHILD to avoid adouble-fork internally
             // since ibus-daemon refuses to start with init as its parent.
-            const [success_, pid] = GLib.spawn_async(
+            const pid = Shell.util_spawn_async(
                 null, cmdLine, env,
-                GLib.SpawnFlags.SEARCH_PATH | GLib.SpawnFlags.DO_NOT_REAP_CHILD,
-                () => {
-                    try {
-                        global.context.restore_rlimit_nofile();
-                    } catch (err) {
-                    }
-                }
-            );
+                GLib.SpawnFlags.SEARCH_PATH | GLib.SpawnFlags.DO_NOT_REAP_CHILD);
             GLib.child_watch_add(
                 GLib.PRIORITY_DEFAULT,
                 pid,
diff --git a/js/misc/util.js b/js/misc/util.js
index 38d0bed..04d09ae 100644
--- a/js/misc/util.js
+++ b/js/misc/util.js
@@ -135,16 +135,9 @@ export function trySpawn(argv) {
 
     try {
         const launchContext = global.create_app_launch_context(0, -1);
-        [success_, pid] = GLib.spawn_async(
+        pid = Shell.util_spawn_async(
             null, argv, launchContext.get_environment(),
-            GLib.SpawnFlags.SEARCH_PATH | GLib.SpawnFlags.DO_NOT_REAP_CHILD,
-            () => {
-                try {
-                    global.context.restore_rlimit_nofile();
-                } catch (err) {
-                }
-            }
-        );
+            GLib.SpawnFlags.SEARCH_PATH | GLib.SpawnFlags.DO_NOT_REAP_CHILD);
     } catch (err) {
         /* Rewrite the error in case of ENOENT */
         if (err.matches(GLib.SpawnError, GLib.SpawnError.NOENT)) {
diff --git a/js/ui/components/networkAgent.js b/js/ui/components/networkAgent.js
index aad9ecf..e665c43 100644
--- a/js/ui/components/networkAgent.js
+++ b/js/ui/components/networkAgent.js
@@ -455,18 +455,12 @@ class VPNRequestHandler extends Signals.EventEmitter {
 
         try {
             const launchContext = global.create_app_launch_context(0, -1);
-            let [success_, pid, stdin, stdout, stderr] =
-                GLib.spawn_async_with_pipes(
+            let [pid, stdin, stdout, stderr] =
+                Shell.util_spawn_async_with_pipes(
                     null, /* pwd */
                     argv,
                     launchContext.get_environment(),
-                    GLib.SpawnFlags.DO_NOT_REAP_CHILD,
-                    () => {
-                        try {
-                            global.context.restore_rlimit_nofile();
-                        } catch (err) {
-                        }
-                    });
+                    GLib.SpawnFlags.DO_NOT_REAP_CHILD);
 
             this._childPid = pid;
             this._stdin = new Gio.UnixOutputStream({fd: stdin, close_fd: true});
